tests/boardfarm_plugins/boardfarm_prplmesh/tests/link_metric_query.py

```diff
@@ -0,0 +1,102 @@
+import time
+from collections import Counter
+from boardfarm.exceptions import SkipTest
+from .prplmesh_base_test import PrplMeshBaseTest
+from capi import tlv
+from opts import debug
+
+
+class LinkMetricQuery(PrplMeshBaseTest):
+    """Check if an agent can report the links it formed with other devices properly
+
+    Devices used in test setup:
+    AP1 - Agent1 [DUT]
+    AP2 - Agent2
+    GW - Controller
+
+    Given:
+    Bi-directional link is formed between AP1 and GW
+    Bi-directional link is formed between AP2 and AP1
+    When:
+    GW controller is instructed; "Link metric query" (0x0005) is sent for "All neighbors" (0) and
+    for "Both Tx and Rx link metrics" (2) from GW controller to AP1 agent
+    Then:
+    Agent must return a "Link metric response" (0x0006) which contains:
+    - A "1905 transmitter link metric" TLV with AP1's local interface MAC address
+    and GW's interface MAC address
+    - A "1905 receiver link metric" TLV with AP1's local interface MAC address
+    and GW's interface MAC address
+    - A "1905 transmitter link metric" TLV with AP1's local interface MAC address
+    and AP2's interface MAC address
+    - A "1905 receiver link metric" TLV with AP1's local interface MAC address
+    and AP2's interface MAC address
+
+    Link's media type (ethernet, wifi), packet stats, RSSI (for wireless rx links) and phy rate
+    (for wireless tx links) are not checked in this test.
+    """
+    def runTest(self):
+        # skip this test if one of the components does not exist in setup
+        try:
+            agent1 = self.dev.DUT.agent_entity
+            agent2 = self.dev.lan2.agent_entity
+            controller = self.dev.lan.controller_entity
+        except AttributeError as ae:
+            raise SkipTest(ae)
+
+        sniffer = self.dev.DUT.wired_sniffer
+        sniffer.start(self.__class__.__name__ + "-" + self.dev.DUT.name)
+
+        mid = controller.ucc_socket.dev_send_1905(agent1.mac, 0x0005,
+                                                  tlv(0x08, 0x0002, "0x00 0x02"))
+        time.sleep(1)
+
+        query = self.check_cmdu_type_single("link metric query", 0x0005,
+                                            controller.mac, agent1.mac,
+                                            mid)
+        query_tlv = self.check_cmdu_has_tlv_single(query, 0x08)
+
+        debug("Checking query type and queried metrics are correct")
+        assert int(query_tlv.link_metric_query_type) == 0x00, "Query type is not 'All neighbors'"
```

Comment:
Why don't you use safe_check_obj_attribute any more?

Reply:
One reason is in test flows we separated exceptions as AssertionErrors which mean that DUT failed and product code should be debugged and other exceptions which meant that tests failed and test itself must be debugged. We can't do that in a straight-forward way in boardfarm. (I really didn't want to leave the ability to distinguish whether the code or the test fails :( ) This undermines the value of safe_check_obj_attribute().

Secondly, boardfarm is designed to fail tests when an exception is raised and if we fail on multiple errors as we do in test_flows.py I think it will be hard to communicate failure reasons with boardfarm. For example when there is a test which checks four different TLV cases and we fail two of them we will have to create an exception which combines the two errors and raises it.

Another reason is safe_check_obj has limits, you have to serialize both sides of the assert statement and it hurts readability.